lib/rtcpeerconnection/firefox.js

```diff
@@ -209,29 +123,50 @@ Object.defineProperty(FirefoxRTCPeerConnection.prototype, 'peerIdentity', {
   })
 });
 
-// NOTE(mmalavalli): Firefox does not support RTCPeerConnection#removeStream(),
-// and calling it will screw up the RTCPeerConnection's internal state. So
-// for now, we don't do anything when it is called.
-// Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=842455
-//
-// UPDATE(mmalavalli): We tried to implement removeStream() using
-// RTCPeerConnection#removeTrack(), but ran into a Firefox bug in the
-// Firefox <--> Chrome interop case, which is mentioned below.
-// Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1133874
-FirefoxRTCPeerConnection.prototype.removeStream = function removeStream() {};
+// NOTE(mmalavalli): Because we are not delegating to the native
+// RTCPeerConnection#removeTrack(), we have to manually maintain a list of added
+// tracks. So we disable the delegation to the native RTCPeerConnection#addTrack()
+// for now.
+FirefoxRTCPeerConnection.prototype.addTrack = function addTrack() {
+  var args = [].slice.call(arguments);
+  var track = args[0];
+  var sender = this._senders.get(track.id);
+  if (sender && sender.track === track) {
+    throw new Error('Cannot add MediaStreamTrack [' + track.id + ', '
+      + track.kind + ']: RTCPeerConnection already has it');
+  }
+  sender = getActiveSenders(this._peerConnection).get(track.id)
+    || this._peerConnection.addTrack.apply(this._peerConnection, args);
 
-// NOTE(mmalavalli): We implement addStream() by calling
-// RTCPeerConnection#addTrack() on each of the MediaStreamTracks in the
-// given MediaStream that have not already been added.
-FirefoxRTCPeerConnection.prototype.addStream = function addStream(mediaStream) {
-  var localTracks = this._peerConnection.getSenders().map(function(sender) {
-    return sender.track;
-  });
-  var newLocalTracks = util.difference(mediaStream.getTracks(), localTracks);
+  this._senders.set(track.id, sender);
+  return sender;
+};
+
+// NOTE(mmalavalli): RTCPeerConnection#removeTrack() has a bug in the
+// Firefox <--> Chrome interop case, which is mentioned below. So we disable
+// its delegation for now.
+// Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1133874
+FirefoxRTCPeerConnection.prototype.removeTrack = function removeTrack(sender) {
+  var track = sender.track;
+  if (!track) {
+    return;
+  }
+  if (this._isClosed) {
+    throw new Error('Cannot remove MediaStreamTrack [' + track.id +
+      ', ' + track.kind + ']: RTCPeerConnection is closed');
+  }
+  sender = this._senders.get(track.id);
+  if (sender && sender.track === track) {
```

Comment:
Should we let the key itself be the MediaStreamTrack? Do we gain anything by using the MediaStreamTrack's ID?